sys/entropy_source/adc_noise/adc_noise.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2020 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @ingroup     sys_entropy_source_adc
+ *
+ * @{
+ * @file
+ *
+ * @author      Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include "periph/adc.h"
+#include "entropy_source/adc_noise.h"
+#include "entropy_source/entropy_source.h"
+
+entropy_source_tests_rep_t adc_state_rep;
+entropy_source_tests_prop_t adc_state_prop;
+
+static int _get_sample(uint8_t *out)
+{
+    int ret = ENTROPY_SOURCE_OK;
+    uint8_t byte = 0;
+
+    for (unsigned bit = 0; bit < 8; bit++) {
+        int sample = adc_sample(ENTROPY_SOURCE_ADC_LINE, CONFIG_ENTROPY_SOURCE_ADC_RES);
+        if (sample < 0) {
+            /* Resolution is not applicable */
+            return ENTROPY_SOURCE_ERR_CONFIG;
+        }
+        /* use LSB of each ADC sample and shift it to build a value */
+        byte |= (uint8_t)(sample & 0x01) << bit;
+    }
+    /* copy generated byte to out */
+    *out = byte;
+
+    if (IS_ACTIVE(CONFIG_ENTROPY_SOURCE_ADC_HEALTH_TEST)) {
+        ret = entropy_source_test(&adc_state_rep, &adc_state_prop, byte);
+    }
+
+    return ret;
+}
```

Comment:
I wasn't aware of #13248, thanks for the pointer and I agree that the entropy of ADC samples heavily depends on hardware resources, which is why certain a priori validation is necessary. 

> when sampling with the maximum sample rate of an ADC, the signal is stable for many successive samples

Are you arguing for a (configurable) sampling interval? 

> the LSB might oscillate at seemingly random patterns

Many things might happen when talking about random processes (that's what one is looking for I guess) and this should not disqualify a module like this. A developer should, however, do the best to give a realistic and conservative estimation about the provided entropy over some time and finally there is basic runtime testing to indicate rough deficits.

The ADC is just one example for an entropy source. When talking about entropy for cryptographic use, a  more advanced module should manage multiple sources and the ADC can be one. Since the ADC does provide entropy, why should we waste it.

Reply:
> Are you arguing for a (configurable) sampling interval?

I only wanted to share my observations. But I also believe that limiting the sample rate could increase the entropy per sample.